File: 07/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const mod = 998244353

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Предвычисляем факториалы и обратные факториалы
	maxN := 400001
	fact := precomputeFactorials(maxN)
	invFact := precomputeInvFactorials(fact, maxN)

	// Читаем количество тестов
	line, _ := reader.ReadString('\n')
	T, _ := strconv.Atoi(strings.TrimSpace(line))

	// Обрабатываем тесты
	for i := 0; i < T; i++ {
		line, _ = reader.ReadString('\n')
		parts := strings.Fields(strings.TrimSpace(line))
		n, _ := strconv.Atoi(parts[0])
		s, _ := strconv.Atoi(parts[1])

		result := solve(n, s, fact, invFact)
		writer.WriteString(fmt.Sprintf("%d\n", result))
	}
}

// solve считает количество замечательных массивов
// Формула: answer(n, s) = (n+1)! × C(s, n)
func solve(n, s int, fact, invFact []int) int {
	if n > s {
		return 0
	}

	// C(s, n) = s! / (n! * (s-n)!)
	c := comb(s, n, fact, invFact)

	// (n+1)! × C(s, n)
	return fact[n+1] * c % mod
}

// comb вычисляет C(n, k) по модулю mod
func comb(n, k int, fact, invFact []int) int {
	if k < 0 || k > n {
		return 0
	}
	return fact[n] * invFact[k] % mod * invFact[n-k] % mod
}

// precomputeFactorials предвычисляет факториалы до n
func precomputeFactorials(n int) []int {
	fact := make([]int, n+1)
	fact[0] = 1
	for i := 1; i <= n; i++ {
		fact[i] = fact[i-1] * i % mod
	}
	return fact
}

// precomputeInvFactorials предвычисляет обратные факториалы
func precomputeInvFactorials(fact []int, n int) []int {
	invFact := make([]int, n+1)
	invFact[n] = modPow(fact[n], mod-2)
	for i := n; i > 0; i-- {
		invFact[i-1] = invFact[i] * i % mod
	}
	return invFact
}

// modPow вычисляет a^b mod mod
func modPow(a, b int) int {
	result := 1
	a %= mod
	for b > 0 {
		if b&1 == 1 {
			result = result * a % mod
		}
		a = a * a % mod
		b >>= 1
	}
	return result
}

File: 07/main.rs
use std::io::{self, BufRead, BufWriter, Write};

const MOD: i64 = 998244353;

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Предвычисляем факториалы и обратные факториалы
    let max_n = 400001;
    let fact = precompute_factorials(max_n);
    let inv_fact = precompute_inv_factorials(&fact, max_n);

    // Читаем количество тестов
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let t: usize = line.trim().parse().unwrap();

    // Обрабатываем тесты
    for _i in 0..t {
        line.clear();
        reader.read_line(&mut line).unwrap();
        let parts: Vec<&str> = line.trim().split_whitespace().collect();
        let n: usize = parts[0].parse().unwrap();
        let s: usize = parts[1].parse().unwrap();

        let result = solve(n, s, &fact, &inv_fact);
        writeln!(writer, "{}", result).unwrap();
    }
    writer.flush().unwrap();
}

// solve считает количество замечательных массивов
// Формула: answer(n, s) = (n+1)! × C(s, n)
fn solve(n: usize, s: usize, fact: &[i64], inv_fact: &[i64]) -> i64 {
    if n > s {
        return 0;
    }

    // C(s, n) = s! / (n! * (s-n)!)
    let c = comb(s, n, fact, inv_fact);

    // (n+1)! × C(s, n)
    fact[n + 1] * c % MOD
}

// comb вычисляет C(n, k) по модулю mod
fn comb(n: usize, k: usize, fact: &[i64], inv_fact: &[i64]) -> i64 {
    if k > n {
        return 0;
    }
    fact[n] * inv_fact[k] % MOD * inv_fact[n - k] % MOD
}

// precompute_factorials предвычисляет факториалы до n
fn precompute_factorials(n: usize) -> Vec<i64> {
    let mut fact = vec![1i64; n + 1];
    for i in 1..=n {
        fact[i] = fact[i - 1] * (i as i64) % MOD;
    }
    fact
}

// precompute_inv_factorials предвычисляет обратные факториалы
fn precompute_inv_factorials(fact: &[i64], n: usize) -> Vec<i64> {
    let mut inv_fact = vec![0i64; n + 1];
    inv_fact[n] = mod_pow(fact[n], MOD - 2);
    for i in (1..=n).rev() {
        inv_fact[i - 1] = inv_fact[i] * (i as i64) % MOD;
    }
    inv_fact
}

// mod_pow вычисляет a^b mod mod
fn mod_pow(mut a: i64, mut b: i64) -> i64 {
    let mut result = 1i64;
    a %= MOD;
    while b > 0 {
        if b & 1 == 1 {
            result = result * a % MOD;
        }
        a = a * a % MOD;
        b >>= 1;
    }
    result
}







